src/snowflake/snowpark/modin/plugin/_internal/index.py

```diff
@@ -0,0 +1,243 @@
+#
+# Copyright (c) 2012-2024 Snowflake Computing Inc. All rights reserved.
+#
+from __future__ import annotations
+
+from typing import Any, Callable, Hashable, Iterator, Literal
+
+import numpy as np
+import pandas as native_pd
+from pandas._typing import ArrayLike, DtypeObj, NaPosition, Self
+from pandas.core.arrays import ExtensionArray
+from pandas.core.indexes.frozen import FrozenList
+
+
+# modin index class, wrapper to native pandas for now
+class Index:
+    """
+    Immutable sequence used for indexing and alignment.
+
+    The basic object storing axis labels for all pandas objects.
+
+    .. versionchanged:: 2.0.0
+
+       Index can hold all numpy numeric dtypes (except float16). Previously only
+       int64/uint64/float64 dtypes were accepted.
+
+    Parameters
+    ----------
+    data : array-like (1-dimensional)
+    dtype : str, numpy.dtype, or ExtensionDtype, optional
+        Data type for the output Index. If not specified, this will be
+        inferred from `data`.
+        See the :ref:`user guide <basics.dtypes>` for more usages.
+    copy : bool, default False
+        Copy input data.
+    name : object
+        Name to be stored in the index.
+    tupleize_cols : bool (default: True)
+        When True, attempt to create a MultiIndex if possible.
+
+    Notes
+    -----
+    An Index instance can **only** contain hashable objects.
+    An Index instance *can not* hold numpy float16 dtype.
+
+    Examples
+    --------
+    >>> pd.Index([1, 2, 3])
+    Index([1, 2, 3], dtype='int64')
+
+    >>> pd.Index(list('abc'))
+    Index(['a', 'b', 'c'], dtype='object')
+
+    >>> pd.Index([1, 2, 3], dtype="uint8")
+    Index([1, 2, 3], dtype='uint8')
+    """
+
+    # same fields as native pandas index constructor
+    def __init__(
+        self,
+        data: Any = None,
+        dtype: Any = None,
+        copy: bool = False,
+        name: Any = None,
+        tupleize_cols: bool = True,
+    ) -> None:
+        from snowflake.snowpark.modin.plugin.compiler.snowflake_query_compiler import (
+            SnowflakeQueryCompiler,
+        )
+
+        # TODO: SNOW-1359041: Switch to lazy index implementation
+        if isinstance(data, native_pd.Index):
+            self._index = data
+        elif isinstance(data, Index):
+            self._index = data.to_pandas()
+        elif isinstance(data, SnowflakeQueryCompiler):
```

Comment:
where do we use this?

Reply:
this is used [here](https://github.com/snowflakedb/snowpark-python/blob/8e3063de303418d0b5dde094207e4d1d604d1179/src/snowflake/snowpark/modin/plugin/compiler/snowflake_query_compiler.py#L1401). It will also be very useful when we need to transition to lazy evaluation